docs(guide/environments): make guide on environments a bit more thorough (#95)
embark-framework/embark-site: source/docs/environments.md

title: Understanding Environments
---

> Embark comes with the concept of "Environments", which make it very easy to switch between different configurations of various parts of our applications that alter how the application is being built and deployed. For example, during development we probably prefer deploying our Smart Contracts on a local blockchain or testnet, until we're sure the code is ready for prime time and can be deployed in a production environment.

Reviewer: ```suggestion
Embark comes with the concept of "Environments", which make it easy to switch between different configurations for various parts of our application, altering how the application is built and deployed. For example, during development we probably prefer to deploy our Smart Contracts to a local blockchain or testnet. When we're sure the code is ready for prime time, it can be deployed to a production environment.
```

In this guide we'll discuss how to take advantage of environments within various configuration files of our Embark application.

## Default environment

> In our guide on [application structures](structure.html) in Embark, we've talked about that every component of our decentralized application, such as IPFS as a storage solution and Geth as a blockchain client, can be configured using a dedicated configuration file. We'll dive more into what each configuration looks like in our guides on [configuring Smart Contracts](contracts_configuration.html), [configuring decentralized storages](storage_configuration.html) and [configuring communication channels](messages_configuration.html), but for now we focus on the concept of **default environments**.

Reviewer: ```suggestion
In our guide covering [application structures](structure.html), we saw that every component of our decentralized application, such as IPFS and Geth, can be configured using a dedicated configuration file. We'll dive more into what each configuration looks like in our guides on [configuring Smart Contracts](contracts_configuration.html), [configuring decentralized storages](storage_configuration.html) and [configuring communication channels](messages_configuration.html). For now, we'll focus on the concept of **default environments**.
```

Environments can be defined as part of a configuration file for a dedicated service or component of our application. We can introduce as many environments as we like. Important is, however, that `default` is a special environment that can be extended with other environments.

Let's take a look at the `config/contracts.js` file that we've created in the [Quickstart](quick_start.html):
<pre><code class="javascript">module.exports = {
  default: {
    deployment: {
      host: "localhost",
      port: 8546,
      type: "ws"
    },
    dappConnection: [
      "$WEB3",
      "ws://localhost:8546",
      "http://localhost:8545"
    ],
    gas: "auto",
    contracts: {
      SimpleStorage: {
        args: [100]
      }
    }
  }
}</code></pre>

Don't get too overwhelmed by all the different options and what mean for now. We'll discuss those in-depth in [configuring Smart Contracts](contracts_configuration.html). The important part here is that `contracts.js` exports a configuration object that comes with a `default` configuration. This configuration is the default environment and can be overwritten or extended with any other environment. 

> If we would execute `$ embark run`, Embark will use the `default` configuration to deploy our application's Smart Contracts.

Reviewer 2: ```suggestion
If we execute `$ embark run`, Embark will use the `default` configuration to deploy our application's Smart Contracts.
```
